File: code/src/prediction_detection_metrics.m
close all
clear all
%% ================= USER INPUT =================
% Ground truth folder containing labeled masks
rootPath = 'F:\Lab\MouseMusclePOGLUT1\results\detectionMetrics';

gtFolder   = fullfile(rootPath, 'imagesTest','GT emi_vicky combined');

% Folder containing predicted masks
% predFolder = fullfile(rootPath, 'imagesTest','modelV1');
namePredImages = 'modelV1';
predFolder = fullfile(rootPath, 'imagesTest',namePredImages);



distThresh = 10;       % Distance threshold (pixels) to consider a centroid matched
outExcel   = fullfile(rootPath, ['detection_evaluation_' namePredImages '_autobinary_160326.xlsx']); % Output Excel file
imgExt     = '*.tif'; % Image file extension
binaryThr  = 250;     % Threshold to binarize predicted masks

%%Min nuclei size
micronsPerPixel = 0.5055443;
minArea_um2 = 10; % minimum nuclei size in um^2
minArea_pixels = minArea_um2 / (micronsPerPixel^2);

%% ==============================================

% List all ground truth files
gtFiles = dir(fullfile(gtFolder, imgExt));

nImages = length(gtFiles); % Total number of images

% Initialize arrays to store per-image metrics
imageNames = string({gtFiles.name})'; % Image file names
TPs = zeros(nImages,1);              % True positives
FPs = zeros(nImages,1);              % False positives
FNs = zeros(nImages,1);              % False negatives
GTs = zeros(nImages,1);              % Number of ground-truth objects
PRs = zeros(nImages,1);              % Number of predicted objects
Precisions = zeros(nImages,1);       % Precision per image
Recalls = zeros(nImages,1);          % Recall per image
F1s = zeros(nImages,1);              % F1-score per image

%% ================= MAIN LOOP =================
for i = 1:nImages
    
    % --- Load images ---
    gt  = imread(fullfile(gtFolder, gtFiles(i).name)); % Ground truth
    pr  = imread(fullfile(predFolder, gtFiles(i).name)); % Prediction
    if ~isa(pr,'uint8')
        BW = uint8(pr*255); % Scale prediction to 0-255 if needed
    else
        BW=pr;
    end
    % --- Binarize images ---
    gt = gt > 0; % Ground truth binary mask
    % Predicted mask: threshold + fill holes + morphological close
    % pr = imclose(bwmorph(BW >= binaryThr,'fill'), strel('disk',1));
    pr = imclose(bwmorph(imbinarize(BW),'fill'), strel('disk',1));


    % Remove small nuclei
    pr = bwareaopen(pr, ceil(minArea_pixels)); % remove small objects

    % imwrite(uint8(pr*255),fullfile(rootPath, 'imagesTest',gtFiles(i).name))

    % --- Label connected components ---
    gtL = bwlabel(gt); % GT labeled objects
    prL = bwlabel(pr); % Predicted labeled objects

    % --- Extract centroids ---
    gtProps = regionprops(gtL, 'Centroid'); % GT centroids
    prProps = regionprops(prL, 'Centroid'); % Predicted centroids

    nGT = numel(gtProps); % Number of GT objects
    nPR = numel(prProps); % Number of predicted objects

    GTs(i) = nGT; % Store number of GT objects
    PRs(i) = nPR; % Store number of predicted objects

    % --- Edge cases: zero objects ---
    if nGT == 0 && nPR == 0
        % Nothing in GT or prediction → perfect
        TPs(i) = 0; FPs(i) = 0; FNs(i) = 0;
        Precisions(i) = 1; Recalls(i) = 1; F1s(i) = 1;
        continue
    end

    if nGT == 0
        % No GT objects → all predictions are false positives
        TPs(i) = 0; FPs(i) = nPR; FNs(i) = 0;
        Precisions(i) = 0; Recalls(i) = 1; F1s(i) = 0;
        continue
    end

    if nPR == 0
        % No predicted objects → all GT are false negatives
        TPs(i) = 0; FPs(i) = 0; FNs(i) = nGT;
        Precisions(i) = 1; Recalls(i) = 0; F1s(i) = 0;
        continue
    end

    % --- Build centroid arrays ---
    gtC = cat(1, gtProps.Centroid); % nGT x 2 array
    prC = cat(1, prProps.Centroid); % nPR x 2 array

    % --- Compute distance matrix between GT and predicted centroids ---
    D = pdist2(gtC, prC); % nGT x nPR

    matchedGT = false(nGT,1); % Track which GT are matched
    matchedPR = false(nPR,1); % Track which predictions are matched

    % --- Greedy one-to-one matching ---
    TP_count = 0;
    while true
        [minD, idx] = min(D(:)); % Find closest pair
        if minD > distThresh
            break % No pair within threshold
        end
        [g,p] = ind2sub(size(D), idx); % Get GT index g and predicted index p

        % Mark as matched
        TP_count = TP_count + 1;
        matchedGT(g) = true;
        matchedPR(p) = true;

        % Remove matched row and column from distance matrix
        D(g,:) = inf;
        D(:,p) = inf;
    end

    % --- Store per-image results ---
    TPs(i) = TP_count;
    FNs(i) = sum(~matchedGT);
    FPs(i) = sum(~matchedPR);

    % Compute Precision / Recall / F1 per image
    Precisions(i) = TPs(i) / (TPs(i) + FPs(i) + eps);
    Recalls(i) = TPs(i) / (TPs(i) + FNs(i) + eps);
    F1s(i) = 2 * Precisions(i) * Recalls(i) / (Precisions(i) + Recalls(i) + eps);
end

%% ================= CREATE TABLE =================
timestamp = datetime('now');

% Per-image table
T = table(imageNames, TPs, FPs, FNs, GTs, PRs, Precisions, Recalls, F1s, ...
    repmat(distThresh,nImages,1), ...
    'VariableNames', {'ImageName','TP','FP','FN','Total_GT','Total_Pred','Precision','Recall','F1','DistanceThreshold_px'});

% --- Add a final row with averages across all images ---
avgT = table( ...
    "Average", ...                  % ImageName
    mean(TPs), ...                  % Mean TP
    mean(FPs), ...                  % Mean FP
    mean(FNs), ...                  % Mean FN
    mean(GTs), ...                  % Mean Total_GT
    mean(PRs), ...                  % Mean Total_Pred
    mean(Precisions), ...           % Mean Precision
    mean(Recalls), ...              % Mean Recall
    mean(F1s), ...                  % Mean F1
    distThresh, ...                 % DistanceThreshold_px
    'VariableNames', T.Properties.VariableNames);

T = [T; avgT]; % Append average row

%% ================= SAVE TO EXCEL =================
if isfile(outExcel)
    % Read existing table
    Told = readtable(outExcel);

    % Ensure matching types for safe concatenation
    if ~isa(Told.ImageName,'string'), Told.ImageName = string(Told.ImageName); end
    numericCols = {'TP','FP','FN','Total_GT','Total_Pred','Precision','Recall','F1','DistanceThreshold_px'};
    for k = 1:length(numericCols)
        col = numericCols{k};
        if ~isa(Told.(col),'double'), Told.(col) = double(Told.(col)); end
    end

    % Concatenate old and new tables
    T = [Told; T];
end

% Write table to Excel
writetable(T, outExcel);

%% ================= REPORT =================
fprintf('\n=== Detection results ===\n');
fprintf('Number of images: %d\n', nImages);
fprintf('Saved per-image metrics and average to: %s\n', outExcel);

File: code/src/prediction_detection_metrics.py
import os
import glob
import math
import numpy as np
import pandas as pd
from datetime import datetime
from skimage import io, measure, morphology, filters
from scipy import ndimage
from scipy.spatial.distance import cdist

# %% ================= USER INPUT =================
# Ground truth folder containing labeled masks
root_path = r'F:\Lab\MouseMusclePOGLUT1\results\detectionMetrics'

gt_folder = os.path.join(root_path, 'imagesTest', 'GT emi_vicky combined')

# Folder containing predicted masks
name_pred_images = 'modelV1'
pred_folder = os.path.join(root_path, 'imagesTest', name_pred_images)

dist_thresh = 10       # Distance threshold (pixels) to consider a centroid matched
out_excel = os.path.join(root_path, f'detection_evaluation_{name_pred_images}_autobinary_160326.xlsx')
img_ext = '*.tif'      # Image file extension
binary_thr = 250       # Threshold to binarize predicted masks (unused in active logic, matching MATLAB)

# Min nuclei size
microns_per_pixel = 0.5055443
min_area_um2 = 10 # minimum nuclei size in um^2
min_area_pixels = math.ceil(min_area_um2 / (microns_per_pixel**2))

# %% ==============================================

# List all ground truth files
search_pattern = os.path.join(gt_folder, img_ext)
gt_files = glob.glob(search_pattern)

n_images = len(gt_files) # Total number of images
eps = np.finfo(float).eps # Machine epsilon to prevent division by zero

# Initialize lists to store per-image metrics (lists are more Pythonic for building DataFrames than pre-allocating zeros)
image_names = []
TPs, FPs, FNs, GTs, PRs, Precisions, Recalls, F1s = [], [], [], [], [], [], [], []

# %% ================= MAIN LOOP =================
for img_path in gt_files:
    file_name = os.path.basename(img_path)
    image_names.append(file_name)
    
    # --- Load images ---
    gt = io.imread(img_path)
    pr_path = os.path.join(pred_folder, file_name)
    pr = io.imread(pr_path)
    
    if pr.dtype != np.uint8:
        BW = (pr * 255).astype(np.uint8)
    else:
        BW = pr
        
    # --- Binarize images ---
    gt_bin = gt > 0
    
    # Predicted mask: threshold (Otsu) + fill holes + morphological close
    # Handle the case where the image might be completely black/constant (Otsu would fail)
    if BW.min() == BW.max():
        pr_bin = BW > 0
    else:
        thresh = filters.threshold_otsu(BW)
        pr_bin = BW > thresh
        
    # Fill holes
    pr_bin = ndimage.binary_fill_holes(pr_bin)
    
    # Morphological close with disk of radius 1
    disk_1 = morphology.disk(1)
    pr_bin = morphology.binary_closing(pr_bin, disk_1)
    
    # Remove small nuclei
    pr_bin = morphology.remove_small_objects(pr_bin, min_size=min_area_pixels)
    
    # --- Label connected components ---
    # connectivity=2 in 2D is equivalent to 8-connectivity in MATLAB
    gt_L = measure.label(gt_bin, connectivity=2)
    pr_L = measure.label(pr_bin, connectivity=2)
    
    # --- Extract centroids ---
    gt_props = measure.regionprops(gt_L)
    pr_props = measure.regionprops(pr_L)
    
    n_GT = len(gt_props)
    n_PR = len(pr_props)
    
    GTs.append(n_GT)
    PRs.append(n_PR)
    
    # --- Edge cases: zero objects ---
    if n_GT == 0 and n_PR == 0:
        TPs.append(0); FPs.append(0); FNs.append(0)
        Precisions.append(1.0); Recalls.append(1.0); F1s.append(1.0)
        continue
        
    if n_GT == 0:
        TPs.append(0); FPs.append(n_PR); FNs.append(0)
        Precisions.append(0.0); Recalls.append(1.0); F1s.append(0.0)
        continue
        
    if n_PR == 0:
        TPs.append(0); FPs.append(0); FNs.append(n_GT)
        Precisions.append(1.0); Recalls.append(0.0); F1s.append(0.0)
        continue
        
    # --- Build centroid arrays ---
    # skimage centroids are (row, col) representing (y, x). Order doesn't matter for distance.
    gt_C = np.array([prop.centroid for prop in gt_props]) # n_GT x 2
    pr_C = np.array([prop.centroid for prop in pr_props]) # n_PR x 2
    
    # --- Compute distance matrix between GT and predicted centroids ---
    D = cdist(gt_C, pr_C) # n_GT x n_PR matrix
    
    matched_GT = np.zeros(n_GT, dtype=bool)
    matched_PR = np.zeros(n_PR, dtype=bool)
    
    # --- Greedy one-to-one matching ---
    TP_count = 0
    while True:
        min_D = np.min(D)
        if min_D > dist_thresh:
            break
            
        # Get 2D indices of the minimum value
        g, p = np.unravel_index(np.argmin(D), D.shape)
        
        # Mark as matched
        TP_count += 1
        matched_GT[g] = True
        matched_PR[p] = True
        
        # Remove matched row and column from distance matrix by setting to infinity
        D[g, :] = np.inf
        D[:, p] = np.inf
        
    # --- Store per-image results ---
    TPs.append(TP_count)
    FNs.append(np.sum(~matched_GT))
    FPs.append(np.sum(~matched_PR))
    
    # Compute Precision / Recall / F1 per image
    p_val = TP_count / (TP_count + FPs[-1] + eps)
    r_val = TP_count / (TP_count + FNs[-1] + eps)
    f1_val = 2 * p_val * r_val / (p_val + r_val + eps)
    
    Precisions.append(p_val)
    Recalls.append(r_val)
    F1s.append(f1_val)

# %% ================= CREATE TABLE =================

# Per-image table
df = pd.DataFrame({
    'ImageName': image_names,
    'TP': TPs,
    'FP': FPs,
    'FN': FNs,
    'Total_GT': GTs,
    'Total_Pred': PRs,
    'Precision': Precisions,
    'Recall': Recalls,
    'F1': F1s,
    'DistanceThreshold_px': [dist_thresh] * n_images
})

# --- Add a final row with averages across all images ---
avg_row = pd.DataFrame({
    'ImageName': ["Average"],
    'TP': [np.mean(TPs)],
    'FP': [np.mean(FPs)],
    'FN': [np.mean(FNs)],
    'Total_GT': [np.mean(GTs)],
    'Total_Pred': [np.mean(PRs)],
    'Precision': [np.mean(Precisions)],
    'Recall': [np.mean(Recalls)],
    'F1': [np.mean(F1s)],
    'DistanceThreshold_px': [dist_thresh]
})

df = pd.concat([df, avg_row], ignore_index=True)

# %% ================= SAVE TO EXCEL =================
if os.path.isfile(out_excel):
    # Read existing table
    df_old = pd.read_excel(out_excel)
    
    # Concatenate old and new tables
    df = pd.concat([df_old, df], ignore_index=True)

# Write table to Excel (index=False prevents pandas from writing row numbers)
df.to_excel(out_excel, index=False)

# %% ================= REPORT =================
print('\n=== Detection results ===')
print(f'Number of images: {n_images}')
print(f'Saved per-image metrics and average to: {out_excel}')